docs/tutorials/demo_best_practices.py

```diff
@@ -0,0 +1,271 @@
+#     language: python
+#     name: python3
+# ---
+
+# %% [markdown]
+"""
+# Data Versioning Best Practices
+
+This notebook will guide you through the best practices for versioning data in a data science project.
+We assume you have lakeFS and lakeFS-spec setup. If you need help with it, look into the other parts of the docs or the other tutorial which is more focused on the setup.
+
+We will explain the following best practices for data versioning in this example:
+- Define a data repository
+- Follow a branching strategy that ensures data integrity on the `main` branch, e.g. by running tests on feature branch datasets.
+- Utilize reusable and tested functions for data transformation.
+- Use commits to save checkpoints and merge branches for atomic changes.
+- Keep naming (of branches and commits) consistent, concise, and unique.
+- Use descriptive naming where it matters .
+
+For this demo project, we aim to build a weather predictor using data from a public API.
+This simulates the dynamics within a real world scenario where we continuously collect more data.
+"""
+# %%
+import json
+import tempfile
+import urllib.request
+from pathlib import Path
+
+import lakefs
+import pandas as pd
+import sklearn
+import sklearn.model_selection
+
+import lakefs_spec
+
+# %% [markdown]
+"""
+The cell below contains a helper function to obtain the data. It is otherwise not relevant to this demonstration.
+"""
+
+
+# %%
+def _maybe_urlretrieve(url: str, filename: str) -> str:
+    # Avoid API rate limit errors by downloading to a fixed local location
+    destination = Path(tempfile.gettempdir()) / "lakefs-spec-tutorials" / filename
+    destination.parent.mkdir(exist_ok=True, parents=True)
+    if destination.exists():
+        return str(destination)
+
+    outfile, _ = urllib.request.urlretrieve(url, str(destination))
+    return outfile
+
+
+outfile = _maybe_urlretrieve(
+    "https://archive-api.open-meteo.com/v1/archive?latitude=52.52&longitude=13.41&start_date=2010-01-01&end_date=2010-12-31&hourly=temperature_2m,relativehumidity_2m,rain,pressure_msl,surface_pressure,cloudcover,cloudcover_low,cloudcover_mid,cloudcover_high,windspeed_10m,windspeed_100m,winddirection_10m,winddirection_100m",
+    "weather-2010.json",
+)
+# %% [markdown]
+"""
+## Define a data repository
+
+We got the data for the year 2010. That should be enough for initial prototyping.
+Later, however, we want to use more data. Since our dataset will be evolving, we implement data version control.
+This ensure the reproducibility of our experiments, enables collaboration with colleagues, and ensures our dynamic dataset to stay a valuable asset.
+
+To set up our versioning, we need to decide on a versioning tool, set up a repository, and define which data is considered in scope and should be versioned and which is not.
+In this case, we want to version the weather data using lakeFS.
+Let us now set up a lakeFS repository using lakeFS-spec, our filesystem implementation for lakeFS.
```

Comment:
Is this worth mentioning? I feel like we can just assume the choice goes to lakeFS everywhere, otherwise we wouldn't be doing this.